### ISSUE TITLE
error: #extension directive is not allowed in the middle of a shader

### ISSUE THREAD

**Issue by tarceri** [yuzu-emu/yuzu#2523]
There is a bug in one of the generated shaders. The bug is triggered on Mesa based linus OpenGL drivers:

 error: #extension directive is not allowed in the middle of a shader

Some other drivers allow extension to be defined mid shader but this is invalid according to the OpenGL spec. You can see the start of the problem shader below:

#define EMULATION_UBO_BINDING 4
#define CBUF_BINDING_1 5
#define CBUF_BINDING_4 6
#define CBUF_BINDING_12 7
layout (lines) in;           <--- This needs to be after #extension
#define MAX_VERTEX_INPUT 2
#extension GL_ARB_separate_shader_objects : enable

Downstream bug report: https://bugs.freedesktop.org/show_bug.cgi?id=110472

**Comment by ReinUsesLisp** [comment]
Thanks for narrowing down this issue. The next canary release should have this fixed.

I'm surprised that some proprietary drivers even allow this.

**Comment by toccata10** [comment]
Still have troubles even with #2524: the system doesn't lock up anymore, but after a few minutes (2-3 minutes), the screen goes black, sound is continuing, but I have to log to a terminal to reboot.
It happens on Arms, SuperMarioOdyssey (maybe others). Superstreetfighters, puyopuyo tetris, they all play fine.
Gpu: r9 fury, mesa 19.0.5, yuzu canaty 2329.
Also tried allow_glsl_extension_directive_midshader=true
Same result (black screen on 3d games after a few minutes).

**Comment by tarceri** [comment]
@toccata10 are you able to try a newer Mesa version (ideally git master)? Either way the issue should be tracked differently from this bug.

**Comment by toccata10** [comment]
Unfortunately not: I had too many problems in the past switching to unstable versions (and I also need my pc for my job). I'm on Arch, so when mesa 19.1 is released, it'll be quickly available.

**Comment by ReinUsesLisp** [comment]
@toccata10 Please open a separate issue related to the crashes. This particular issue is fixed on master.

**Comment by e88z4** [comment]
@tarceri I have tried it with latest mesa master. It didn't fix the original issue which is the ring gfx timeout.

@ReinUsesLisp I have an old issue which was closed. Should the closed issue be re-opened? If you want a new issue to be opened, I can create one.
https://github.com/yuzu-emu/yuzu/issues/1865

This issue is related to Mesa bug tracker below:
https://bugs.freedesktop.org/show_bug.cgi?id=109022
https://bugs.freedesktop.org/show_bug.cgi?id=110472
